netebpfext/user/kernel_um.h

```diff
@@ -0,0 +1,278 @@
+
+void
+ExWaitForRundownProtectionRelease(_Inout_ EX_RUNDOWN_REF* rundown_ref);
+
+BOOLEAN
+ExAcquireRundownProtection(_Inout_ EX_RUNDOWN_REF* rundown_ref);
+
+void
+ExReleaseRundownProtection(_Inout_ EX_RUNDOWN_REF* rundown_ref);
+
+_Acquires_exclusive_lock_(push_lock->lock) void ExAcquirePushLockExclusiveEx(
+    _Inout_ _Requires_lock_not_held_(*_Curr_) _Acquires_lock_(*_Curr_) EX_PUSH_LOCK* push_lock,
+    _In_ unsigned long flags);
+
+_Acquires_shared_lock_(push_lock->lock) void ExAcquirePushLockSharedEx(
+    _Inout_ _Requires_lock_not_held_(*_Curr_) _Acquires_lock_(*_Curr_) EX_PUSH_LOCK* push_lock,
+    _In_ unsigned long flags);
+
+_Releases_exclusive_lock_(push_lock->lock) void ExReleasePushLockExclusiveEx(
+    _Inout_ _Requires_lock_held_(*_Curr_) _Releases_lock_(*_Curr_) EX_PUSH_LOCK* push_lock, _In_ unsigned long flags);
+
+_Releases_shared_lock_(push_lock->lock) void ExReleasePushLockSharedEx(
+    _Inout_ _Requires_lock_held_(*_Curr_) _Releases_lock_(*_Curr_) EX_PUSH_LOCK* push_lock, _In_ unsigned long flags);
+
+void*
+ExAllocatePoolUninitialized(_In_ POOL_TYPE pool_type, _In_ size_t number_of_bytes, _In_ unsigned long tag);
+
+void
+ExFreePool(void* p);
+
+void
+ExInitializePushLock(_Out_ EX_PUSH_LOCK* push_lock);
+
+void
+FatalListEntryError(_In_ void* p1, _In_ void* p2, _In_ void* p3);
+
+MDL*
+IoAllocateMdl(
+    _In_opt_ __drv_aliasesMem void* virtual_address,
+    _In_ unsigned long length,
+    _In_ BOOLEAN secondary_buffer,
+    _In_ BOOLEAN charge_quota,
+    _Inout_opt_ IRP* irp);
+
+PIO_WORKITEM
+IoAllocateWorkItem(_In_ DEVICE_OBJECT* device_object);
+
+void
+IoQueueWorkItem(
+    _Inout_ __drv_aliasesMem IO_WORKITEM* io_work_item,
+    _In_ IO_WORKITEM_ROUTINE* worker_routine,
+    _In_ WORK_QUEUE_TYPE queue_type,
+    _In_opt_ __drv_aliasesMem void* context);
+
+void
+IoFreeWorkItem(_In_ __drv_freesMem(Mem) PIO_WORKITEM io_work_item);
+
+void
+IoFreeMdl(MDL* mdl);
+
+void
+KeEnterCriticalRegion(void);
+
+void
+KeLeaveCriticalRegion(void);
+
+void
+KeInitializeSpinLock(_Out_ PKSPIN_LOCK spin_lock);
+
+_Requires_lock_not_held_(*spin_lock) _Acquires_lock_(*spin_lock) _IRQL_requires_max_(DISPATCH_LEVEL) KIRQL
+    KeAcquireSpinLockRaiseToDpc(_Inout_ PKSPIN_LOCK spin_lock);
+
+_Requires_lock_held_(*spin_lock) _Releases_lock_(*spin_lock) _IRQL_requires_(DISPATCH_LEVEL) void KeReleaseSpinLock(
+    _Inout_ PKSPIN_LOCK spin_lock, _In_ _IRQL_restores_ KIRQL new_irql);
+
+void
+MmBuildMdlForNonPagedPool(_Inout_ MDL* memory_descriptor_list);
+
+unsigned long
+MmGetMdlByteCount(_In_ MDL* mdl);
+
+void*
+MmGetSystemAddressForMdlSafe(
+    _Inout_ MDL* mdl,
+    _In_ unsigned long page_priority // MM_PAGE_PRIORITY logically OR'd with MdlMapping*
+);
+
+NTSTATUS
+RtlULongAdd(
+    _In_ unsigned long augend,
+    _In_ unsigned long addend,
+    _Out_ _Deref_out_range_(==, augend + addend) unsigned long* result);
+
+// Inline functions
+_Must_inspect_result_ BOOLEAN CFORCEINLINE
+IsListEmpty(_In_ const LIST_ENTRY* list_head)
```

Comment:
If `LIST_ENTRY` struct was obtained from some header, why couldnt the inline functions too be obtained from there?

Reply:
LIST_ENTRY and IsListEmpty are in different header files.
LIST_ENTRY is in winnt.h
IsListEmpty is in wdm.h

wdm.h conflicts with many of the user mode headers and can't be included, hence the duplication of the functions we need.